src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Unix/CertificateAssetDownloader.cs

```diff
@@ -0,0 +1,127 @@
+        {
+            byte[] data = DownloadAsset(uri, ref remainingDownloadTime);
+
+            if (data == null)
+            {
+                return null;
+            }
+
+            try
+            {
+                return new X509Certificate2(data);
+            }
+            catch (CryptographicException)
+            {
+                return null;
+            }
+        }
+
+        internal static unsafe byte[] DownloadAsset(string uri, ref TimeSpan remainingDownloadTime)
+        {
+            if (remainingDownloadTime <= TimeSpan.Zero)
+            {
+                return null;
+            }
+
+            List<byte[]> dataPieces = new List<byte[]>();
+
+            using (Interop.libcurl.SafeCurlHandle curlHandle = Interop.libcurl.curl_easy_init())
+            {
+                GCHandle gcHandle = GCHandle.Alloc(dataPieces);
+                IntPtr dataHandlePtr = GCHandle.ToIntPtr(gcHandle);
+
+                try
+                {
+                    Interop.libcurl.curl_easy_setopt(curlHandle, Interop.libcurl.CURLoption.CURLOPT_URL, uri);
+                    Interop.libcurl.curl_easy_setopt(curlHandle, Interop.libcurl.CURLoption.CURLOPT_WRITEDATA, dataHandlePtr);
+                    Interop.libcurl.curl_easy_setopt(curlHandle, Interop.libcurl.CURLoption.CURLOPT_WRITEFUNCTION, s_writeCallback);
+
+                    Stopwatch stopwatch = Stopwatch.StartNew();
+                    int res = Interop.libcurl.curl_easy_perform(curlHandle);
+                    stopwatch.Stop();
+
+                    // TimeSpan.Zero isn't a worrisome value on the subtraction, it only
+                    // means "no limit" on the original input.
+                    remainingDownloadTime -= stopwatch.Elapsed;
+
+                    if (res != Interop.libcurl.CURLcode.CURLE_OK)
+                    {
+                        return null;
+                    }
+                }
+                finally
+                {
+                    gcHandle.Free();
+                }
+            }
+
+            if (dataPieces.Count == 0)
+            {
+                return null;
+            }
+
+            if (dataPieces.Count == 1)
+            {
+                return dataPieces[0];
+            }
+
+            int dataLen = 0;
+
+            for (int i = 0; i < dataPieces.Count; i++)
+            {
+                dataLen += dataPieces[i].Length;
+            }
+
+            byte[] data = new byte[dataLen];
+            int offset = 0;
+
+            for (int i = 0; i < dataPieces.Count; i++)
+            {
+                byte[] piece = dataPieces[i];
+
+                Buffer.BlockCopy(piece, 0, data, offset, piece.Length);
+                offset += piece.Length;
+            }
+
+            return data;
+        }
+
+        private static unsafe ulong CurlWriteCallback(byte* buffer, ulong size, ulong nitems, IntPtr context)
+        {
+            ulong totalSize = size * nitems;
```

Comment:
When @nguerrera's "shim everything" project gets to libcurl this will probably have a more managed-focused solution.  Until then, this is the signature that we need to match libcurl's callback.